daffodil-core/src/main/scala/org/apache/daffodil/grammar/ElementBaseGrammarMixin.scala

```diff
@@ -1368,4 +1371,32 @@ trait ElementBaseGrammarMixin
       mtaBase
     }
 
+  val allowedBitTypes = Set[PrimType](
+    PrimType.Boolean,
+    PrimType.Byte,
+    PrimType.Short,
+    PrimType.Int,
+    PrimType.Long,
+    PrimType.UnsignedByte,
+    PrimType.UnsignedShort,
+    PrimType.UnsignedInt,
+    PrimType.UnsignedLong,
+  )
+  val allowedBitTypesText = allowedBitTypes.map("xs:" + _.toString).toList.sorted.mkString(", ")
+
+  private def assertLengthUnits(elem: ElementBase = context): Unit = {
+    elem.lengthUnits match {
+      case LengthUnits.Bits if elem.representation == Representation.Binary =>
+        elem.optPrimType match {
+          case Some(primType) =>
+            if (!allowedBitTypes.contains(primType))
+              if (tunable.allowBigIntegerBits)
+                SDW(WarnID.DeprecatedBigIntegerBits, s"In a future release, lengthUnits='bits' will only be supported for the following types: $allowedBitTypesText")
+              else
+                SDE("lengthUnits='bits' is only supported for the following types: $allowedBitTypesText")
```

Comment:
Since `elem` can be `context` or `repElement`, I wonder if we need to call SDE/SDW on `elem` instead of `this` since the SDE/SDW methods include the calling object's source location in the printed message?

Reply:
Updated to call `elem.{SDE,SDW}`